source/common/upstream/upstream_impl.cc

```diff
@@ -919,9 +919,19 @@ ClusterImplBase::ClusterImplBase(
 
   auto socket_matcher = std::make_unique<TransportSocketMatcherImpl>(
       cluster.transport_socket_matches(), factory_context, socket_factory, *stats_scope);
-  info_ = std::make_unique<ClusterInfoImpl>(cluster, factory_context.clusterManager().bindConfig(),
-                                            runtime, std::move(socket_matcher),
-                                            std::move(stats_scope), added_via_api, factory_context);
+  auto& dispatcher = factory_context.dispatcher();
+  info_ = std::shared_ptr<const ClusterInfoImpl>(
+      new ClusterInfoImpl(cluster, factory_context.clusterManager().bindConfig(), runtime,
+                          std::move(socket_matcher), std::move(stats_scope), added_via_api,
+                          factory_context),
+      [&dispatcher](const ClusterInfoImpl* self) {
+        const auto name = self->name();
+        ENVOY_LOG(debug, "Schedule destroy cluster info {}", self->name());
+        dispatcher.movePost([raii_cluster = std::shared_ptr<const ClusterInfoImpl>(self)]() {
```

Comment:
It took me a while to understand this movePost function, but I think I understand now the risk associated with a post callback that holds a shared_ptr executing before the state in the thread that scheduled the post full unrolls.

Use of raw pointers doesn't work if for some reason the callback execution is skipped due to shutdown of the main thread's dispatcher.

Reply:
> Use of raw pointers doesn't work if for some reason the callback execution is skipped due to shutdown of the main thread's dispatcher.

Yes, I tried also repeated drain post_callbacks_ but more failures are introduced. 
The latest effort is to introduce preShutdown(), so that we can destroy movePosted object. 

Another question is, do we really need to destroy the cluster info when the server is shutting down? My hunch is that the destruction is nice to have but not mandatory.

Reply:
I wonder if using a raw pointer here and sequencing things so that worker threads are always destroyed before the main thread dispatcher stops would solve some of these issues.

Leaking ClusterInfoImpl objects during shutdown is not great for tests, but it may be unavoidable to move in that direction depending on how we decide to solve the dispatcher shutdown problem.  We need to come up with a plan for dispatcher shutdown.

Reply:
I know it's an immature decision to introduce preShutdown() in the dispatcher. 
However, it's probably the easy way to fix hundreds of test failure, even though we are confident the test failures are low signal. 

I tried several times to remove preShutdown but no luck. Do you think a preShutdown is acceptable? If not,  how about `preShutdownForTest` to indicate no further production logic should be put within?

Reply:
Adding an explicit shutdown sequence for the dispatcher seems like a potential long term strategy.

Reply:
We are almost there!